src/lib/testing/random.ts

```diff
@@ -364,7 +644,7 @@ function int(min: number, max: number): Random<number> {
     create: () => () => {
       // 25% of test cases are special numbers
```

Comment:
I considered that.. The reason I ended up not doing it is that as these random generators are combined into higher-level structures, it seems much more valuable to randomly combine special values, like 0, with other values. Also, we use random values for things like array sizes which determine the structure of the generated output. If, say, 0 would always be the first value that is drawn, then the _first sample_ of anything would contain just empty arrays, null / undefined and zeroes; instead of sprinkling zeroes, nulls and empty arrays throughout the samples, like it is done now.

I wouldn't worry about the less than 100% probability. In practice, we run these tests so often (on every push to a PR) that we're guaranteed to catch failing tests. (I would worry much more about missing tests than about ones that are not catching bugs in 1/100 runs)

Take this test for a point in case: https://github.com/o1-labs/snarkyjs/blob/d6b7995aa81070c1dcc78eb751ae70578ddeb97a/src/lib/testing/testing.unit-test.ts#L19-L24
I checks that the `nat` generator, when used twice to sample tuples `(x,y)`, generates at least one sample where `x === y`. In our current system where special values are created with 25% probability, and there are only a few special values, this should work most of the time but fail with low probability. However, it hasn't failed a single time in CI in about a month since it has been added, so this slight imperfection hasn't been a problem in practice

Reply:
Consistency is extremely important for the developer and it concerns me something could sneak by.  Without updating your generic framework, I would strongly suggest that the zero/(special case) tests can always be hard-coded to always happen.

Reply:
(as a separate kind of test)

Reply:
That's a good suggestion! It's actually possible to add hard-coded values (which are always tested first) to any random generator by using `Random.withHardCoded(generator, ...<list of hard-coded values>)`. I use this in few places currently but will check if I can use it in more places to add important test cases that we want to test with 100% certainty